pipefuser/modules/patch_parallel/patch_parallel_conv2d.py

```diff
@@ -0,0 +1,337 @@
+        return ((start + stride - 1) // stride) * stride
+
+    def get_world_size_and_rank(self):
+        rank = 0
+        world_size = 1
+        if dist.is_available() and dist.is_initialized():
+            rank = dist.get_rank()
+            world_size = dist.get_world_size()
+        return world_size, rank
+
+    def get_kernel_size_and_stride(self):
+        # formatting kernel_size and stride
+        if isinstance(self.kernel_size, int):
+            kernel_size_h, kernel_size_w = self.kernel_size, self.kernel_size
+        elif isinstance(self.kernel_size, tuple):
+            kernel_size_h, kernel_size_w = self.kernel_size
+        else:
+            raise ValueError(
+                f"kernel_size should be int or tuple, type:{type(self.kernel_size)}"
+            )
+
+        if isinstance(self.stride, int):
+            stride_h, stride_w = self.stride, self.stride
+        elif isinstance(self.stride, tuple):
+            stride_h, stride_w = self.stride
+        else:
+            raise ValueError(
+                f"stride should be int or tuple, type: {type(self.stride)}")
+        return kernel_size_h, kernel_size_w, stride_h, stride_w
+
+    def _all_gather(self, result: Tensor):
+        b, c, h, w = result.shape
+        world_size, rank = self.get_world_size_and_rank()
+        cuda = torch.device(f"cuda:{rank}")
+        if world_size == 1:
+            return result
+        tensor_list = [
+            torch.zeros(1, dtype=torch.int64, device=cuda)
+            for _ in range(world_size)
+        ]
+        dist.all_gather(tensor_list,
+                        torch.tensor([h], dtype=torch.int64, device=cuda))
+        segment_list = [t.item() for t in tensor_list]
+        max_h = max(segment_list)
+        padding = [0, 0, 0, max_h - h]
+        result = F.pad(result, padding, mode="constant")
+        result_list = [
+            torch.zeros([b, c, max_h, w], dtype=result.dtype, device=cuda)
+            for _ in range(world_size)
+        ]
+        dist.all_gather(result_list, result)
+        post_result_list = [
+            res[:, :, :segment_list[i], :] for i, res in enumerate(result_list)
+        ]
+        return torch.cat(post_result_list, dim=2)
+
+    def _conv_forward(self, input: Tensor, weight: Tensor,
+                      bias: Optional[Tensor]) -> Tensor:
+        inner_input = input
+        world_size, rank = self.get_world_size_and_rank()
+
+        def adjust_padding(padding: list, rank: int):
+            if rank == 0:
+                padding[-1] = 0
+            elif rank == world_size - 1:
+                padding[-2] = 0
+            else:
+                padding[-2:] = [0, 0]
+
+        if self.padding_mode != 'zeros':
+            copy_padding = self._reversed_padding_repeated_twice[:]
+            if world_size > 1:
+                adjust_padding(copy_padding, rank)
+            inner_input = F.pad(input, copy_padding, mode=self.padding_mode)
```

Comment:
It seems to be padding twice under the case `self.padding_mode != 'zeros' &&  self.padding != 0`. But it have no impact on the models we currently support.

Reply:
There is also padding in the conv op, which I brought out because not all sides of a patch are needed to pad.  The padding parameter of  conv is set 0, such as, `F.conv2d(inner_input, weight, bias, self.stride, 0,
                            self.dilation, self.groups)`,

Reply:
Yeah. But I mean the duplicated F.pad in lines 143 and 161. If self.padding_mode != 'zeros' && self.padding != 0, F.pad(input, copy_padding, mode=self.padding_mode) in line 143 will be executed and then F.pad(inner_input, padding_list, mode="constant") in line 161. Is that an problem?

Reply:
All right, I got it, and I have fixed it.